File: Game/Action.py
class Action:

    def __init__(self, start_pos: tuple[int, int], end_pos: tuple[int, int]):
        self.start_pos = start_pos
        self.end_pos = end_pos

    def __str__(self):
        return f'{self.start_pos} -> {self.end_pos}'

    def __eq__(self, other):
        return isinstance(other, Action) and self.start_pos == other.start_pos and self.end_pos == other.end_pos

File: java_chess/src/Action.java
import java.util.Objects;

public class Action {
  public Pos start_pos;
  public Pos end_pos;

  public Action(Pos start_pos, Pos end_pos) {
    this.start_pos = start_pos;
    this.end_pos = end_pos;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Action action = (Action) o;
    return start_pos.equals(action.start_pos) && end_pos.equals(action.end_pos);
  }

  @Override
  public String toString() {
    return "" + start_pos + " -> " + end_pos;
  }

  @Override
  public int hashCode() {
    return Objects.hash(start_pos, end_pos);
  }
}